Improve email mirror and outgoing SMTP documentation.

timabbott opened the pull request.

smarx commented:
*Automated message from Dropbox CLA bot*

@timabbott, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
@karamcnair can you review at least the email mirror part of this?  Thanks!

karamcnair commented:
Will do. I have never reviewed a PR before (especially not a diff on github) and am about to head into a session, so I will probably not get to it until this evening - is that ok?

timabbott commented:
Yeah no problem!

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -45,27 +45,30 @@
-# likely need to configure that account as "less secure" here:
-# https://support.google.com/accounts/answer/6010255.
+# Configure the outgoing SMTP server below. You will need to complete
+# the following steps:
```
This may be overkill, but it would be useful for me to see a list of the categories of outgoing emails that are sent (missed messages, are there others?) so as to know why it's important to configure this (side question - is outgoing SMTP actually _required_ for a self-hosted Zulip install or is it only required if they want to support missed messages & whatever the other emails are)?

timabbott replied:
It's required for password authentication in addition to things like missed message emails.  

So if you're using Google or LDAP auth and thus aren't using Zulip passwords and you don't need missed message notifications you could do without it.

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -45,27 +45,30 @@
-# you're having issues, you may want to search for documentation on
-# using your email provider with Django.
+# (1) Fill out the outgoing email sending configuration below.  You
+# can skip sending emails entirely by commenting out EMAIL_HOST.
```
Oh, this answers my question about whether sending emails is required. I'd personally bump that up outside a list item b/c the set of 3 steps is irrelevant if I don't want to send steps & comment out EMAIL_HOST instead.

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -148,39 +151,47 @@
 # The email gateway provides, for each stream, an email address that
```
For me, this text was a bit hard to read because it assumes that I have a mental concept of an 'email gateway' that I don't know about yet. From a newbie setup perspective, some wording like this would work better for me:

"Zulip supports emailing messages to be posted to streams by creating an email address for each stream to receive messages at. The format for the email address is defined by the EMAIL_GATEWAY_PATTERN configuration item. This is called 'email mirroring' in Zulip. 

[aside: why is the word 'local' in "There are two ways to make use of local email mirroring"? Is there a non-local way of doing email mirroring?]

"There are two ways to configure email mirroring in Zulip: 
...

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -148,39 +151,47 @@
-
-# The following options are relevant if you're using mail polling.
+# The local delivery configuration is preferred for production because
+# it doesn't have delay, while the polling mechanism is preferred for
```
This is again, probably overkill, but for me (a noob) reading this, I read "is preferred for development" as meaning that it's preferred by the project maintainers in some way when I think what is meant is "most people doing development prefer setting up an IMAP mailbox and polling it" (the former sounding like a prescription, the latter sounding like advice/guidance)

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -148,39 +151,47 @@
+# for auto-generated stream emails.  The %s will be replaced with a
+# unique token, and the resulting email must be delivered to the
+# EMAIL_GATEWAY_IMAP_FOLDER of the EMAIL_GATEWAY_LOGIN account below,
+# or piped in to the email-mirror management command as indicated
```
I didn't see what was indicated above: "piped in to the email-mirror management command as indicated above" Did I read badly? (it happens)

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -148,39 +151,47 @@
+# or piped in to the email-mirror management command as indicated
+# above.
 #
+# E.g. %s@example.com (local delivery) or username+%s@example.com (polling)
```
A clarification question: username+%s@example.com is basically required for a mailbox at username@example.com right? In order to get the mail delivered to username@example.com

Is %s@example.com the only option for local delivery or would username+%s@example.com also work for local delivery? I actually don't know the answer to this but I suspect the latter would work too. Personally, this is where some examples would have been useful, e.g. "If you have set up an IMAP mailbox at zulipadmin@mydomain.com, then using zulipadmin+%s@mydomain.com will generate stream email addresses that look like 'zulipadmin+streamidentifier+978699987@mydomain.com' and missed message reply email addresses that look like 'zulipadmin+089656647@mydomain.com'. If you set up local delivery, the same format would work, but you could also use "%s@mydomain.com" and the addresses would look like 'streamidentifier +978699987@mydomain.com' and '089656647@mydomain.com'"

I understand that that is somewhat implementation dependent, but this would have been much clearer to me :)

karamcnair commented on zproject/local_settings_template.py:
```diff
@@ -148,39 +151,47 @@
+# form username+%s@example.com (which is what you will set as
+# EMAIL_GATEWAY_PATTERN); your email provider should deliver those
+# emails to the username@example.com inbox.  Then you run in a cron
+# job `./manage.py email-mirror` (see puppet/zulip/files/cron.d/email-mirror),
```
Is this what was moved from above that is referred to in 'as indicated above'?

karamcnair commented:
I added some very wordy comments :)

timabbott commented:
OK @karamcnair I pushed some new commits (which I plan to squash into the originals before merging but left separate to make it easier for you to read) addressing your comments.

One thing I realized while editing this is that the EMAIL_GATEWAY_BOT default should be fine for all installations so we don't need to document that as something the user needs to set, so I just removed it from the file entirely.

karamcnair commented:
Looks great!

timabbott commented:
Merged, thanks for all the feedback @karamcnair!
